src/app/tests/TestCommandInteraction.cpp

```diff
@@ -540,6 +573,27 @@ void TestCommandInteraction::AddInvokeResponseData(nlTestSuite * apSuite, void *
     }
 }
 
+void TestCommandInteraction::FillCurrentInvokeResponseBuffer(nlTestSuite * apSuite, CommandHandler * apCommandHandler,
+                                                             const ConcreteCommandPath & aRequestCommandPath,
+                                                             uint32_t aSizeToLeaveInBuffer)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    err = apCommandHandler->AllocateBuffer();
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+    uint32_t remainingSize = apCommandHandler->mInvokeResponseBuilder.GetWriter()->GetRemainingFreeLength();
+
+    // This number was derived by executing this method with aSizeToLeaveInBuffer = 100 and
+    // subsequently analyzing the remaining size to determine the overhead associated with calling
+    // AddResponseData with `ForcedSizeBuffer`.
+    uint32_t sizeNeededForAddingResponse = 27;
```

Comment:
This won't scale to changes to IM messages. Is there a different way to get the same test going without requiring exactly the right edge bounds? Like, could a bigger estimate be used, and then make the message slightly bigger than max, rather "exactly max", and slightly lower than max, and test both cases do the right thing? Otherwise, is it possible to reuse some primitives/utilities to build a response in a buffer, in such a way that the final buffer's .size() can be used to make this scale to changes to encoding?

This can be done in a follow-up.

Reply:
To test the extreme where we fill the buffer to the edge (to really test rollback). Yes this needs to exact max, this actually found a real bug that I fixed in this PR.

I have an idea of how IM layers can provide the values, but like you said in a follow up